url04f7fcf4e1_peteratt_HDFS_Replication_Simulator_tgz-pJ8-hdfs_replicationsimulator_MainJ8/src/hdfs/replicationsimulator/Simulator.java

```diff
@@ -0,0 +1,275 @@
+package hdfs.replicationsimulator;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Random;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/*
+ * Object used to Initialize the system. Create all the objects and processes.
+ */
+public class Simulator {
+
+	private static Namenode namenode;
+	private static AllDatanode allDatanodes;
+
+	private static Queue<Event> toNamenode;
+	private static Queue<Event> toDatanodes;
+
+	private static List<SimTrace> traceList;
+
+	/*
+	 * Settings (defaults)
+	 */
+	private static int numberofBlocks = 75000;
+	private static int numberofReplicas = 3;
+	private static int numberofDatanodes = 10000;
+	private static int dataNodeCapacity = 320000;
+	private static int bandwidth = 1024;
+	private static int heartbeat = 1000;
+	private static int timeout = 3;
+	private static int blockSize = 64;
+
+	private static List<Event> simulationFailureEvents;
+
+	public static void init(String configFile) {
+		// Read the test.txt file: CHANGE THE NAME!
+		try {
+			FileReader fr = new FileReader(configFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *fr*. Execution paths that do not contain closure statements were detected. Ensure *fr* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).